Makefile

```diff
@@ -30,20 +33,40 @@ define build_ligo
 	$(foreach CVAR,$(3),$(file >>$(TOTAL_FILE),#define $(CVAR)))
 	@echo "#include \"$(notdir $(1))\"" >> $(TOTAL_FILE)
 
-	# ============== Compiling `$(1)` with options `$(3)` ============== #
+	# ============== Compiling Ligo Contract `$(1)` with options `$(3)` ============== #
 	@$(BUILD) $(TOTAL_FILE) main --output-file $(2) || ( rm $(TOTAL_FILE) && exit 1 )
 	@rm $(TOTAL_FILE)
 endef
 
+define build_ligo_storage
+	@mkdir -p $(dir $(2))
+
+	@ #Create a file and put necessary #define pragmas to it first
+	$(eval TOTAL_FILE := $(shell mktemp $(1).total-XXX))
+	$(foreach CVAR,$(3),$(file >>$(TOTAL_FILE),#define $(CVAR)))
+	@echo "#include \"$(notdir $(1))\"" >> $(TOTAL_FILE)
+
+	# ============== Compiling Ligo Storage `$(1)` with options `$(3)` ============== #
+	@$(BUILD_STORAGE) $(TOTAL_FILE) main default_storage --output-file $(2) || ( rm $(TOTAL_FILE) && exit 1 )
+	@rm $(TOTAL_FILE)
+endef
+
 all: \
-	$(OUT)/segmented_cfmm_default.tz
+	$(OUT)/segmented_cfmm_default.tz $(OUT)/storage_default.tz
 
 $(OUT)/segmented_cfmm_default.tz : LIGO_PRAGMAS = DUMMY_PRAGMA1 DUMMY_PRAGMA2
 
 # Generic rule for compiling CFMM contract variations.
 $(OUT)/segmented_cfmm_%.tz: ligo/**
 	$(call build_ligo,ligo/main.mligo,$(OUT)/segmented_cfmm_$*.tz,$(LIGO_PRAGMAS))
 
+
+$(OUT)/storage_default.tz : LIGO_PRAGMAS = DUMMY_PRAGMA1 DUMMY_PRAGMA2
+
+$(OUT)/storage_%.tz: ligo/**
+	$(call build_ligo_storage,ligo/main.mligo,$(OUT)/storage_$*.tz,$(LIGO_PRAGMAS))
+
```

Comment:
Make sense, I've updated this, should be good now.

Reply:
@rinn7e the only addition I would make is to remove these files in the `haskell/test` dir on the `clean` target as well

Reply:
I update the `clean` to remove those files as well.

One problem coming up is that, the compile time code that read contract/storage expect relative path like this:

```
defaultStorage :: Storage
defaultStorage =
  fromVal ($(fetchValue @Storage "../haskell/test/storage_default.tz" "STORAGE_PATH"))
```

This works locally, but won't work in the CI for some reasons: https://buildkite.com/serokell/segmented-cfmm/builds/241#e0c7b80b-0d1f-4f48-a93b-a99d30234584/72-228

As a result, I still have to pass the flag in the Makefile to make the CI passed:
```
test: prepare_lib
	$(MAKE) -C haskell test PACKAGE=segmented-cfmm \
		SEGMENTED_CFMM_PATH=../haskell/test/segmented_cfmm_default.tz \
		STORAGE_PATH=../haskell/test/storage_default.tz
```

Let me know if you're alright with this.

Reply:
It should be possible for CI to copy the contract as well.

Firstly I would change the path in `defaultStorage` to be `test/storage_default.tz` which seems like it should work just as well.

Secondly, there's a way to copy the artifacts from one step to another, in the `pipeline.yaml` you can add a utility "env" before the `steps:` definitions:
```yaml
env:
  FETCH_CONTRACT: >-
    mkdir -p haskell/test/
    && buildkite-agent artifact download ligo-out/storage_default.tz haskell/test/
    && buildkite-agent artifact download ligo-out/segmented_cfmm_default.tz haskell/test/
```
and in the haskell steps you can use this in `commands` by calling `eval "$FETCH_CONTRACT"`, e.g.:
```yaml
  - label: build-haskell
    key: build-haskell
    commands:
      - eval "$FETCH_CONTRACT"
      - nix-build ci.nix -A all-components
```

How does it seem to you? Can you try it out?

Reply:
Seems like the CI only works with path `haskell/test/segmented_cfmm_default.tz`.

I already updated according to your feedback, it manages to download the artifact fine, but still it does not recognized the files: https://buildkite.com/serokell/segmented-cfmm/builds/246#6b59a189-bc36-4695-9fd3-4f68c4b97dda/65-220

Reply:
Ah sorry, my suggestion was misguided, I didn't understand your previous comment correctly and only now I noticed that the step didn't depend on `build-ligo`.

It looks like the building is repeated and the files copied already as part of `packages.segmented-cfmm` definition in `ci.nix`, so everything should work without your last commit and without env vars in the Makefile.

I pushed this change [here](https://github.com/serokell/segmented-cfmm/tree/pasqu4le/tcfmm-12-tmp) to check this and as you can see [here](https://buildkite.com/serokell/segmented-cfmm/builds/252#4a0e6beb-8b72-4119-9fd8-b6e42c6d8519) it seems to work fine :shrug: 

Can you check it and cherry-pick it in your own branch?

Reply:
> It looks like the building is repeated and the files copied already as part of packages.segmented-cfmm definition in ci.nix, so everything should work without your last commit and without env vars in the Makefile.


I see, I've updated `ci.nix` to use `test` dir instead of `haskell/test` dir to match the local build (local build requires `test/<file.tz>` path). Also removed the `eval` path since it is redundant this should be good now, please check this again.

Reply:
@rinn7e  LGTM :ok_hand: